### DISCUSSION TITLE
Do tasks run in parallel or sequential in one executor?

### DISCUSSION THREAD

**Discussion by HaoYang670** [NVIDIA/spark-rapids#5337] [General] [answered]
I am running a query 
`df.withColumn("b", f.col('value').cast("STRING")).agg({'b':'max'}).collect()` in spark-rapids. 
The `df` is stored as parquet format on my desktop, with 1e8 rows and 1 column. Each row is an array[byte] with 50 elements. 
My desktop has 1 CPU (16 cores) and 1 GPU. The spark is in standalone mode with 1 worker (1 executor). The configuration `spark.rapids.sql.concurrentGpuTasks` = `16`
I got the following information from spark UI:
![image](https://user-images.githubusercontent.com/59198230/145352791-46888f15-4419-4958-afe4-73a7f08378b4.png)
![image](https://user-images.githubusercontent.com/59198230/145353191-b6ed4fd1-6f12-4487-a8ac-4acbaf0b1e38.png)

It seems like the 16 tasks does not run in parallel (you can see the min and max time of  `GpuScan`. Also, durations of all tasks make an arithmetic sequence). But I declare that `concurrentGpuTask = 16`?

Do the 16 tasks run in parallel or in sequential?

**Comment by tgravescs** [comment, marked as answer]
the parallelism overall on the executor is first controlled by the number of cores given to the executor, how many does the executor have?  you said the machine has 16 but were all of those allocated to your 1 executor?  Easiest way to check is look at the executors UI page. 
If that had 16 cores that means Spark could have started 16 tasks in parallel on that executor and the spark.rapids.sql.concurrentGpuTasks does control how many of those tasks are allowed to run on the GPU concurrently.  So it should control the max parallelism of tasks run concurrently on the GPU.

more information on this at: https://nvidia.github.io/spark-rapids/docs/tuning-guide.html#number-of-tasks-per-executor

Also note that this is how much could run in parallel its up to spark to schedule and if tasks are so quick to finish sometimes you may not get all running at the same time, you can watch the UI while running the application and see how many are running on the executor at once either by looking at executor page and refreshing or looking at the stages page.

**Comment by jlowe** [comment]
Closing this as answered.  @HaoYang670 please reopen if you have further questions.
